### PR TITLE
Scheduler run_job framework + Reload svr_conns during sched HUP + IFL…

### PR DESCRIPTION
**Pull request by suresh-thelkar**
#### Describe Bug or Feature

This PR consists of the following changes.

- All the run_job related code changes which ensures that scheduler sends the jobs to the right server with the target server's hostname and port. Also need to make code changes ATTR_pbs_server/ATTR_server of Job and Node objects to have the respective pbs_server conf variable value. This includes both Server/Scheduler side changes.
- Fix for intermittent IFL call failures(especially before new server comes up fully if the IFL queries go to new server)
- Made Scheduler code changes **to reload svr_conns table** with the right data. After this I am able to test these changes by modifying PBS_SERVER_INSTANCES in pbs_conf without stopping PBS and verified that now sched reloads all svr_conns table.
- Even after the above changes pbs_add_server script started failing due to couple of race conditions when Server contacting scheduler. Find the root causes for the same and fixed it.
- Handling Misconfiguration at pbs_sched (for example more number of servers getting connected than configured)
-  **Making Server connect to remote Server :** Right now default scheduler does not connect to a different remote server in another complex. This was by design. Now I had to change it for Multi-Server so that two different PBS Complexes still can use single Scheduler. May be we need to rewrite some PTL test cases around this. We can take it up when we do it for mainline.
- Fixed a race condition in Scheduler Failover.

#### Attach Test and Valgrind Logs/Output

I have tested the code changes by setting up multiple servers to the extent possible. Following are the logs for the same.
[MultiSvrTesting_v1.txt](https://github.com/subhasisb/openpbs/files/4888871/MultiSvrTesting_v1.txt)

### CODE REVIEW

**Comment by suresh-thelkar** [comment]
All the test cases from Travis and AppVeyor are passed. OpenSUSE is failing due to the know issue.

**Comment by suresh-thelkar** [comment]
Requesting @nithinj/ @subhasisb  reviews.

File: src/scheduler/fifo.c [hunk 1]
```diff
@@ -1303,14 +1304,73 @@ update_job_can_not_run(int pbs_sd, resource_resv *job, schd_error *err)
...
+
+		extend = malloc(MAX_SVR_ID);
+		if (extend == NULL) {
+			log_event(PBSEVENT_ERROR, PBS_EVENTCLASS_SCHED, LOG_ERR, __func__, "malloc failure");
+			return -1;
+		}
+
+		svr_id_len = strlen(svr_conns[i]->svr_id);
+		extend[svr_id_len] = '\0';
+		strncpy(extend, svr_conns[i]->svr_id, svr_id_len);
```
Comment (nithinj) [inline src/scheduler/fifo.c:1364]: Why not simply do an strdup?
extend is getting freed inside IFL?

File: src/lib/Libutil/misc_utils.c [hunk 2]
```diff
@@ -2214,7 +2217,7 @@ random_srv_conn(svr_conn_t **svr_connections)
 	srand(time(0));
 	ind =  rand() % get_current_servers();
 
-	if (svr_connections[ind]->state == SVR_CONN_STATE_CONNECTED)
+	if ((svr_connections[ind] != NULL) && svr_connections[ind]->state == SVR_CONN_STATE_CONNECTED)
```
Comment (nithinj) [inline src/lib/Libutil/misc_utils.c:2220]: just check for `svr_connections[ind]`
`if (svr_connections[ind] && svr_connections[ind]->state == SVR_CONN_STATE_CONNECTED)`

File: src/scheduler/fifo.c [hunk 3]
```diff
@@ -1303,14 +1304,73 @@ update_job_can_not_run(int pbs_sd, resource_resv *job, schd_error *err)
...
+				return -1;
+			}
+
+			colon_ptr = strstr(tmp_str, ":") ;
+			if (colon_ptr == NULL) {
+				log_event(PBSEVENT_ERROR, PBS_EVENTCLASS_SCHED, LOG_ERR, __func__, "malformed svr_id");
+				return -1;
+			}
+
+			*colon_ptr = '\0';
```
Comment (nithinj) [inline src/scheduler/fifo.c:1342]: instead of duping, you can also restore the colon back

File: src/scheduler/fifo.c (hunk 1 above)
Comment (suresh-thelkar) [reply src/scheduler/fifo.c]: Even strdup() is also not required. I have optimised the code. Please look into the latest commits. With this code change no need to free extend.

File: src/scheduler/fifo.c (hunk 3 above)
Comment (suresh-thelkar) [reply src/scheduler/fifo.c:1335]: I am now restoring the colon.

**Comment by suresh-thelkar** [comment]
@nithinj, I have addressed your comments. Please take a look.

I have also added one more commit which has code changes to fix a race condition that happens when all servers try to connect to Scheduler in parallel.

@subhasisb / @nithinj, Can you please also review this new commit as well.

File: src/include/pbs_ifl.h [hunk 4]
```diff
@@ -158,6 +158,7 @@ extern "C" {
 #define ATTR_state	"job_state"
 #define ATTR_queue	"queue"
 #define ATTR_server	"server"
+#define ATTR_pbs_server	"pbs_server"
```
[the remark this answers is not in the document]
Comment (suresh-thelkar) [reply src/include/pbs_ifl.h]: Yes. I have renamed it to ATTR_server and did all the corresponding changes.

File: src/scheduler/fifo.c [hunk 5]
```diff
@@ -1303,14 +1304,67 @@ update_job_can_not_run(int pbs_sd, resource_resv *job, schd_error *err)
...
+
+			colon_ptr = strstr(svr_conns[i]->svr_id, ":") ;
+			if (colon_ptr == NULL) {
+				log_event(PBSEVENT_ERROR, PBS_EVENTCLASS_SCHED, LOG_ERR, __func__, "malformed svr_id");
+				return -1;
+			}
+
+			*colon_ptr = '\0';
+
+			if (strcmp(svr_conns[i]->svr_id, svr_of_node) == 0) {
```
[the remark this answers is not in the document]
Comment (suresh-thelkar) [reply src/scheduler/fifo.c]: Yes it is in fact going to the server who owns the job. This particular code is to find out the destination server and its port. Please look at the following line of code which takes care of this. 
src_svr_index = get_svr_index(svr_of_job);

File: src/scheduler/fifo.c [hunk 6]
```diff
@@ -1303,14 +1304,67 @@ update_job_can_not_run(int pbs_sd, resource_resv *job, schd_error *err)
...
+		}
+
+
+		for (i = 0; i < num_conf_svrs; i++) {
+			char *colon_ptr;
+
+			if (svr_conns[i] == NULL)
+				continue;
+
+			colon_ptr = strstr(svr_conns[i]->svr_id, ":") ;
```
[the remark this answers is not in the document]
Comment (suresh-thelkar) [reply src/scheduler/fifo.c]: Replaced strstr with strchr.

**Comment by suresh-thelkar** [comment]
@nithinj, Thanks for sign off.
@agrawalravi90, I have implemented all your comments.

**Pull request merged**